sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/state/StateDataSourceV2.scala

```diff
@@ -0,0 +1,164 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.spark.sql.execution.datasources.v2.state
+
+import java.util
+import java.util.UUID
+
+import org.apache.hadoop.conf.Configuration
+import org.apache.hadoop.fs.Path
+
+import org.apache.spark.sql.{AnalysisException, RuntimeConfig, SparkSession}
+import org.apache.spark.sql.connector.catalog.{Table, TableProvider}
+import org.apache.spark.sql.connector.expressions.Transform
+import org.apache.spark.sql.execution.streaming.{CommitLog, OffsetSeqLog, OffsetSeqMetadata}
+import org.apache.spark.sql.execution.streaming.StreamingSymmetricHashJoinHelper.{LeftSide, RightSide}
+import org.apache.spark.sql.execution.streaming.state.{StateSchemaCompatibilityChecker, StateStore, StateStoreConf, StateStoreId, StateStoreProviderId}
+import org.apache.spark.sql.sources.DataSourceRegister
+import org.apache.spark.sql.types.StructType
+import org.apache.spark.sql.util.CaseInsensitiveStringMap
+
+class StateDataSourceV2 extends TableProvider with DataSourceRegister {
+  import StateDataSourceV2._
+
+  private lazy val session: SparkSession = SparkSession.active
+
+  private lazy val hadoopConf: Configuration = session.sessionState.newHadoopConf()
+
+  override def shortName(): String = "statestore"
+
+  override def getTable(
+      schema: StructType,
+      partitioning: Array[Transform],
+      properties: util.Map[String, String]): Table = {
+    val checkpointLocation = Option(properties.get(PARAM_PATH)).orElse {
+      throw new AnalysisException(s"'$PARAM_PATH' must be specified.")
+    }.get
+
+    val resolvedCpLocation = resolvedCheckpointLocation(checkpointLocation)
+
+    val batchId = Option(properties.get(PARAM_BATCH_ID)).map(_.toLong).orElse {
+      Some(getLastCommittedBatch(resolvedCpLocation))
+    }.get
+
+    val operatorId = Option(properties.get(PARAM_OPERATOR_ID)).map(_.toInt)
+      .orElse(Some(0)).get
+
+    val storeName = Option(properties.get(PARAM_STORE_NAME))
+      .getOrElse(StateStoreId.DEFAULT_STORE_NAME)
+
+    val joinSide = Option(properties.get(PARAM_JOIN_SIDE))
+      .map(JoinSideValues.withName).getOrElse(JoinSideValues.none)
+
+    val stateConf = buildStateStoreConf(resolvedCpLocation, batchId)
+
+    val stateCheckpointLocation = new Path(resolvedCpLocation, "state")
+    new StateTable(session, schema, stateCheckpointLocation.toString, batchId, operatorId,
+      storeName, joinSide, stateConf)
+  }
+
+  override def inferSchema(options: CaseInsensitiveStringMap): StructType = {
+    val checkpointLocation = Option(options.get(PARAM_PATH)).orElse {
+      throw new AnalysisException(s"'$PARAM_PATH' must be specified.")
+    }.get
+
+    val resolvedCpLocation = resolvedCheckpointLocation(checkpointLocation)
```

Comment:
Can we combine into a single function for all these confs ?

Seems like we are doing the same operations above ?

```
    val checkpointLocation = Option(properties.get(PARAM_PATH)).orElse {
      throw new AnalysisException(s"'$PARAM_PATH' must be specified.")
    }.get

    val resolvedCpLocation = resolvedCheckpointLocation(checkpointLocation)

    val batchId = Option(properties.get(PARAM_BATCH_ID)).map(_.toLong).orElse {
      Some(getLastCommittedBatch(resolvedCpLocation))
    }.get

    val operatorId = Option(properties.get(PARAM_OPERATOR_ID)).map(_.toInt)
```